### PR TITLE
[12.0] Add module sql_export_excel

### PR DESCRIPTION
**Pull request by florian-dacosta**
This new module extend the sql_export module and add the possibility to extract the data in an excel file.
It also make possible to insert the result of the sql query directly in a excel template file. Which can be usefull when the result of the query is used in other sheets with a lot of calculation for example.

I did a refactore in sql_export to allow to have different format (csv and excel).
And also a very small refactore on sql_request_abstract to allow to add header or not in the excel file.
@legalsylvain @bealdav @aheficent

### CODE REVIEW

File: sql_export_excel/models/sql_export.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+# Copyright 2019 Akretion
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from openerp import api, exceptions, fields, models, _
```
Comment (bealdav) [inline sql_export_excel/models/sql_export.py]:
```suggestion
from odoo import api, exceptions, fields, models, _
```

File: sql_export/models/sql_export.py [hunk 2]
```diff
@@ -48,3 +53,18 @@ def export_sql_query(self):
             'context': self.env.context,
             'nodestroy': True,
         }
+
+    def _get_file_extension(self):
+        self.ensure_one()
+        if self.file_format == 'csv':
+            return 'csv'
+
+    def csv_get_datas_from_query(self, variable_dict):
```
Comment (bealdav) [inline sql_export/models/sql_export.py]:
```suggestion
    def csv_get_data_from_query(self, variable_dict):
```

File: sql_export/wizard/wizard_file.py [hunk 3]
```diff
@@ -68,16 +68,15 @@ def export_sql(self):
...
 
-        # Execute Request
-        res = sql_export._execute_sql_request(
-            params=variable_dict, mode='stdout',
-            copy_options=sql_export.copy_options)
-        if self.sql_export_id.encoding:
-            res = res.decode(self.sql_export_id.encoding)
+        # Call different method depending on file_type since the logic will be
+        # different
+        method_name = '%s_get_datas_from_query' % sql_export.file_format
```
Comment (bealdav) [inline sql_export/wizard/wizard_file.py]:
```suggestion
        method_name = '%s_get_data_from_query' % sql_export.file_format
```

File: sql_export/wizard/wizard_file.py [hunk 4]
```diff
@@ -68,16 +68,15 @@ def export_sql(self):
...
+        datas = getattr(sql_export, method_name)(variable_dict)
```
Comment (bealdav) [inline sql_export/wizard/wizard_file.py]:
```suggestion
        data = getattr(sql_export, method_name)(variable_dict)
```

File: sql_export/wizard/wizard_file.py [hunk 5]
```diff
@@ -68,16 +68,15 @@ def export_sql(self):
...
+        extension = sql_export._get_file_extension()
         self.write({
-            'binary_file': res,
-            'file_name': '%(name)s_%(date)s.csv' % {
-                'name': sql_export.name, 'date': date}
+            'binary_file': datas,
```
Comment (bealdav) [inline sql_export/wizard/wizard_file.py]:
```suggestion
            'binary_file': data,
```

File: sql_export_excel/models/sql_export.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                raise exceptions.ValidationError(
+                    _("The column position can't be less than 1."))
+
+    @api.multi
+    def _get_file_extension(self):
+        self.ensure_one()
+        if self.file_format == 'excel':
+            return 'xlsx'
+        else:
+            return super(SqlExport, self)._get_file_extension()
```
Comment (bealdav) [inline sql_export_excel/models/sql_export.py]:
```suggestion
            return super()._get_file_extension()
```

File: sql_export_excel/tests/test_sql_query_excel.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+try:
+    import openpyxl
+except ImportError:
+    _logger.debug('Can not import openpyxl')
+
+
+class TestExportSqlQueryExcel(TransactionCase):
+
+    def setUp(self):
+        super(TestExportSqlQueryExcel, self).setUp()
```
Comment (bealdav) [inline sql_export_excel/tests/test_sql_query_excel.py]:
```suggestion
        super().setUp()
```

**Review by bealdav — COMMENTED** [review]
suggested

File: sql_export_excel/models/sql_export.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    @api.multi
+    def excel_get_datas_from_query(self, variable_dict):
```
Comment (bealdav) [inline sql_export_excel/models/sql_export.py]:
```suggestion
    def excel_get_data_from_query(self, variable_dict):
```

**Review by bealdav — APPROVED** [review]

**Review by kevinkhao — APPROVED** [review]
Functional tests:
Works as expected, raised errors are understandable for user
LGTM

**Comment by florian-dacosta** [comment]
@dreispt 
I am not sure why the merge failed.
Is it possible to retry? thanks

**Pull request merged**